Assets/MixedRealityToolkit-SDK/Features/Input/MixedRealityInputManager.cs

```diff
@@ -275,37 +280,47 @@ public override void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.Event
                 return;
             }
 
-            GameObject focusedObject = FocusProvider?.GetFocusedObject(baseInputEventData);
+            if (baseInputEventData.InputSource.Pointers == null)
+            {
+                Debug.LogError($"InputSource {baseInputEventData.InputSource.SourceName} doesn't have any registered pointers! Input Sources without pointers should use the GazeProvider's pointer as a default fallback.");
+                return;
+            }
 
-            // Handle modal input if one exists
-            if (modalInputStack.Count > 0)
+            // Get the focused object for each pointer of the event source
+            for (int i = 0; i < baseInputEventData.InputSource.Pointers.Length; i++)
             {
-                GameObject modalInput = modalInputStack.Peek();
+                GameObject focusedObject = FocusProvider?.GetFocusedObject(baseInputEventData.InputSource.Pointers[i]);
 
-                // If there is a focused object in the hierarchy of the modal handler, start the event bubble there
-                if (focusedObject != null && modalInput != null && focusedObject.transform.IsChildOf(modalInput.transform))
+                // Handle modal input if one exists
+                if (modalInputStack.Count > 0)
                 {
-                    if (ExecuteEvents.ExecuteHierarchy(focusedObject, baseInputEventData, eventHandler) && baseInputEventData.used)
+                    GameObject modalInput = modalInputStack.Peek();
```

Comment:
It looks like this will send the same event to the modal stack for every pointer, even though the modal object isn't going to change. Is there a reason for that? Seems like it would lead to duplicate events and event handling.

Reply:
Depends on if the pointer handles the event or not. We still check if the event is used.

If it's not used, then it should go to every handler.

Reply:
Not all handlers will mark the event as used though. Some might want to respond to an event but let it continue down to other objects.

Reply:
If the developer gets duplicate events, then they need to mark the eventData as used.

Reply:
>Previously, this was fine because the modal object would only receive one event. Now, it will receive one for every pointer.

This doesn't handle "modal" input, only handles the currently focused objects. So there is no impact there.  If an object is focused by two different pointers, then it should get both events.

Reply:
This does send events to the modal listeners: https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2920/files#diff-83ff2790c22f67c0d7d4a7a413be9c29R310

I agree about the two different pointers scenario. That's not what I'm talking about.

Reply:
Maybe we should check if the `modalInput` is the same as the focused object?

Thoughts?

Reply:
```suggestion
                    GameObject modalInput = modalInputStack.Peek();

                    if (modalInput != focusedObject) { continue; }
```

Reply:
This code checks to see if the `focusedObject` is a child of `modalInput`, but can probably be removed in this new world with multiple focused objects (unless we want to get all focused objects, do this check for all focused objects, send the modal event as appropriate, then send the focused events): https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2920/files#diff-83ff2790c22f67c0d7d4a7a413be9c29R300

If we take the above suggested change, `modalInput` will never get the event unless it's also the `focusedObject`, which goes against the purpose of the modal input stack.

The event should be sent to the modal object once, probably before the pointers are iterated through.

Reply:
> The event should be sent to the modal object once, probably before the pointers are iterated through.

I agree

Reply:
>If we take the above suggested change, modalInput will never get the event unless it's also the focusedObject, which goes against the purpose of the modal input stack.

Yikes, yeah can't do that. 

> The event should be sent to the modal object once, probably before the pointers are iterated through.

The only way to get the focused object is with a pointer reference, and we could potentially have multiple focused objects.

I'm fairly certain that most (if not all) modal inputs consume the event.

Reply:
>The only way to get the focused object is with a pointer reference, and we could potentially have multiple focused objects.

Yes, I agree. How does this relate to sending the event to the modal input?

Reply:
I've got an idea. One tick.